ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java

```diff
@@ -453,6 +463,56 @@ private void cleanupOldState() {
     localCommands.ifPresent(lc -> lc.processLocalCommandFiles(serviceContext));
   }
 
+  public void cleanupOldStateDirectories(final KsqlConfig configWithApplicationServer) {
+    final String stateDir = configWithApplicationServer.getKsqlStreamConfigProps().getOrDefault(
+            StreamsConfig.STATE_DIR_CONFIG,
+            StreamsConfig.configDef().defaultValues().get(StreamsConfig.STATE_DIR_CONFIG)).toString();
+
+    final Set<String> stateStoreNames =
+        ksqlEngine.getPersistentQueries()
+            .stream()
+            .map(PersistentQueryMetadata::getQueryApplicationId)
+            .collect(Collectors.toSet());
+    try {
+      Files.walkFileTree(Paths.get(stateDir), new SimpleFileVisitor<Path>() {
```

Comment:
@guozhangwang updated this, lmk if it looks good to you

Reply:
Left a few more comments.